structured kernels - error check when structured_delegate is not marked structured

bdhirsh opened the pull request:
Adding a quick error check for structured kernels. Given the below entry in `native_functions.yaml`:
```
 - func: opA(...) -> ...
   structured_delegate: opB
```

If `opB` is not marked as `structured: True`, you end up with nasty compiler errors about a missing class. We can create a better error in the codegen by checking that the delegated-to op is marked `structured: True`.

I think this is the first assertion in the codegen that requires examining multiple native_functions together across NativeFunctionGroup boundaries, so I added a small bit for that.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52227 structured kernels - error check when structured_delegate is not marked structured**

Differential Revision: [D26431105](https://our.internmc.facebook.com/intern/diff/D26431105)

bdhirsh committed:
structured kernels - error check when structured_delegate is not marked structured

[ghstack-poisoned]

ezyang commented on tools/codegen/gen.py:
```diff
@@ -78,8 +78,23 @@ def parse_native_yaml(path: str) -> List[NativeFunction]:
+            assert delegate_func.structured, \
+                f"{f.func.name} is marked as a structured_delegate pointing to " \
+                f"{f.structured_delegate}, but {f.structured_delegate} is not marked as structured. " \
+                f"Consider adding 'structured=True' to the delegated operator"
```
I think you can do the check instead at StructuredNativeFunctions.from_dict and then you don't need to construct the map here

bdhirsh replied:
`StructuredNativeFunctions.from_dict` is only aware of the NativeFunctions within a single NativeFunction group. So if I add it in there, we wouldn't be able to enforce the check across NativeFunctionGroup boundaries. That doesn't really fit with the fact that we allow users to pass in any operator name to `structured_delegate`, right?

cc @bhosmer

bhosmer replied:
One question is whether we really want to allow users to pass in any operator name. Do we have use cases that need that permissiveness? Anyway though given that's what the system currently allows, this seems like the right check.

bdhirsh replied:
Yeah, I was wondering that too- I remember us talking a little about it a little during in a composability sync about difference options for the structured syntax in `native_functions.yaml` and this was settled on. If we ever want to change that, e.g. force `structured_delegate` to always delegate to the `out` variant in the same NativeFunctionGroup, it should be pretty easy to do that.

bdhirsh committed:
Update on "structured kernels - error check when structured_delegate is not marked structured"

Adding a quick error check for structured kernels. Given the below entry in `native_functions.yaml`:
```
 - func: opA(...) -> ...
   structured_delegate: opB
```

If `opB` is not marked as `structured: True`, you end up with nasty compiler errors about a missing class. We can create a better error in the codegen by checking that the delegated-to op is marked `structured: True`.

I think this is the first assertion in the codegen that requires examining multiple native_functions together across NativeFunctionGroup boundaries, so I added a small bit for that.


Differential Revision: [D26431105](https://our.internmc.facebook.com/intern/diff/D26431105)

[ghstack-poisoned]
